### PR TITLE
[Pool Simulator] Implement gyro 2CLP math

### PR DESCRIPTION
**Pull request by luizakp**
This PR adds the Gyro 2CLP math interface

### CODE REVIEW

File: apps/balancer-tools/src/app/poolsimulator/(utils)/index.ts [hunk 1]
```diff
@@ -58,24 +59,25 @@ export async function convertAnalysisDataToAMM(data: AnalysisData) {
...
-    // }
+    case PoolTypeEnum.Gyro2: {
+      return new AMM(
+        new ExtendedGyro2({
+          swapFee: String(data.poolParams?.swapFee),
+          totalShares: String(
+            data.tokens.reduce((acc, token) => acc + token.balance, 0),
+          ),
+          tokens: data.tokens.map((token) => ({
+            address: String(token.symbol), // math use address as key, but we will use symbol because custom token will not have address
```
Comment (ribeirojose) [inline apps/balancer-tools/src/app/poolsimulator/(utils)/index.ts:70]: does it really need to be coerced to String? 🤔

File: apps/balancer-tools/src/contexts/PoolSimulatorContext.tsx [hunk 2]
```diff
@@ -25,7 +26,10 @@ import {
 import { PoolAttribute } from "#/components/SearchPoolForm";
 import { pools } from "#/lib/gql";
 
-export type PoolPairData = MetaStablePoolPairData | GyroEPoolPairData;
+export type PoolPairData =
```
Comment (ribeirojose) [inline apps/balancer-tools/src/contexts/PoolSimulatorContext.tsx:29]: Could be imported from math package, no?

File: packages/math-poolsimulator/src/gyro2/index.test.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+      tokens: poolData.tokens,
+      tokensList: poolData.tokensList,
+      sqrtAlpha: poolData.sqrtAlpha,
+      sqrtBeta: poolData.sqrtBeta,
+    }),
+  );
+
+  const pool = ExtendedGyro2.fromPool(poolData as SubgraphPoolBase);
+  const poolPairData = pool.parsePoolPairData(tokenIn, tokenOut);
+  const amount = 50742;
```
[the remark this answers is not in the document]
Comment (yvesfracari) [reply packages/math-poolsimulator/src/gyro2/index.test.ts:28]: E.g. test the function for amounts equal to 25, 50, 75, 95 percent of the token balance -- for both cases.

Also, we could variate all parameters in a range and test again.

File: packages/math-poolsimulator/src/gyro2/index.test.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  const amountOldBigNumber = bnum(amount);
+  const spotPriceExpected = pool._spotPriceAfterSwapExactTokenInForTokenOut(
+    poolPairData,
```
Comment (yvesfracari) [inline packages/math-poolsimulator/src/gyro2/index.test.ts:31]: Just curious, did we ever tested what happens if we pass 0 to this function?

File: packages/math-poolsimulator/src/gyro2/index.ts [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+      afterFeeMultiplier,
+      virtOut,
+    );
+    const newSpotPrice = GyroHelpersSignedFixedPoint.divDown(
+      numerator,
+      denominator,
+    );
+
+    return newSpotPrice;
+  }
```
Comment (yvesfracari) [inline packages/math-poolsimulator/src/gyro2/index.ts:123]: Makes sense for me but would test how the usage of the `_calculateNewSpotPrice` works passing `inAmount` and `outAmount` as `BigNumber(0)` .

File: packages/math-poolsimulator/src/gyro2/index.ts [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+
+  _spotPrice(poolPairData: Gyro2PoolPairData): OldBigNumber {
```
Comment (yvesfracari) [inline packages/math-poolsimulator/src/gyro2/index.ts:125]: Also seems right for me but would check what happens if we pass 0 as the amount on the `_spotPriceAfterExactTokenInForTokenOut` function.

**Review by yvesfracari — DISMISSED** [review]
LGTM! The test comments that I added were more like an idea about how to solve the frontend errors that we are having (would not work on then on this PR).

File: packages/math-poolsimulator/src/gyro2/index.test.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (luizakp) [reply packages/math-poolsimulator/src/gyro2/index.test.ts:28]: great suggestion

[the remark this answers is not in the document]
Comment (luizakp) [reply packages/math-poolsimulator/src/gyro2/index.test.ts:28]: logged this here: https://linear.app/bleu-llc/issue/BAL-607/test-the-spot-price-for-amounts-equal-to-25-50-75-95-percent-of-the

File: packages/math-poolsimulator/src/gyro2/index.test.ts (hunk 4 above)
Comment (luizakp) [reply packages/math-poolsimulator/src/gyro2/index.test.ts:31]: image
 😭 it works just like our spot price kkkkkk

Comment (luizakp) [reply packages/math-poolsimulator/src/gyro2/index.test.ts:31]: will open a new PR to solve this

File: apps/balancer-tools/package.json [hunk 7]
```diff
@@ -49,7 +49,7 @@
     "inputmask": "^5.0.8",
     "lodash": "^4.17.21",
     "lodash.merge": "^4.6.2",
-    "next": "^13.4.13",
```
Comment (ribeirojose) [inline apps/balancer-tools/package.json:52]: ❌

**Pull request merged**